### PR TITLE
Handle specific branch.io deeplinks

### PR DESCRIPTION
**Pull request by tchayen**
Fixes RNBW-2491

## What changed (plus any additional context for devs)

Due to recent problems with branch.io, we are aiming to switch to aggressive redirect mode. There appears to be two different ways Branch is deeplinking to the app, one works as expected and is handled properly with current code, the other appears to have shape `rainbow://open?_branch_referrer=A&link_click_id=B` and current code fails to decode it (as it generally looses all data URL apart from cryptic `_branch_referrer` query param which turns out to be a series of transformations to the original URL (`encodeURIComponent(base64(gzip(link)))`).

## PoW (screenshots / screen recordings)

No visual changes.

## Dev checklist for QA: what to test

All possible universal link/deeplinks in the app:
- https://rainbow.me/ens_name.eth
- WC connect, signing
- Ethereum transaction links

Of course there's no need to confirm anything, just that expected modal appears in the app.

## Final checklist

- [x] Assigned individual reviewers?
- [x] Added labels?

### CODE REVIEW

**Review by ibrahimtaveras00 — APPROVED** [review]
Deeplinks worked fine on both OS 👍🏽 

Although I didn't have any issues with opening the links, this would pop up for me only on Android: 

 ERROR  Error from Branch: Trouble initializing Branch. Invalid or missing app id, Branch key, or secret. The request was invalid.

Screen Shot 2022-02-09 at 6 29 30 PM

seen here = https://recordit.co/enbrtkwfj0

**Comment by tchayen** [comment]
This issue seems to happen also on other branches 👀 . I think we should investigate that separately.

**Review by osdnk — COMMENTED** [review]
Can we try maybe updating SDK first? We're currently using the version for 2016 and I relying on this param seems to be a bit fragile

File: src/App.js [hunk 1]
```diff
@@ -184,6 +196,23 @@ class App extends Component {
     this.branchListener?.();
   }
 
+  decodeBranchUrl = source => {
```
Comment (osdnk) [inline src/App.js:199]: can we move this (and maybe whole branch logic) to a file?

Comment (osdnk) [reply src/App.js:199]: + I would wrap this into try-catch or add at least a few optionals?

File: src/App.js [hunk 2]
```diff
@@ -136,7 +138,17 @@ class App extends Component {
...
       if (params['+non_branch_link']) {
         const nonBranchUrl = params['+non_branch_link'];
-        this.handleOpenLinkingURL(nonBranchUrl);
+
+        // This happens when branch.io redirects user to the app in the
+        // aggressive redirect mode with a confirmation modal in Safari.
+        // Those URLs have shape rainbow://open?_branch_referrer=A&link_click_id=B
+        if (nonBranchUrl.startsWith('rainbow://open')) {
+          this.handleOpenLinkingURL(this.decodeBranchUrl(nonBranchUrl));
+          // This happens when user taps Rainbow universal link managed by branch.io
```
[the remark this answers is not in the document]
Comment (tchayen) [reply src/App.js:147]: Done ✅

**Comment by osdnk** [comment]
Okay, I was thinking that in this case we can move all branch-related logic to a single file 😅. In was always a bit weird to handle this directly in App.js

But this is your call

**Review by skylarbarrera — APPROVED** [review]
lgtm, making a ticket for one android branch issue

**Review by osdnk — APPROVED** [review]
Lfg

**Pull request merged**